### ISSUE TITLE
x/build: run a Windows XP or Server 2003 builder

### ISSUE THREAD

**Issue by bradfitz** [golang/go#10267]
We should run a Windows Server 2003 builder.  (not just WS2008/WS2012 on GCE)

In order of difficulty:

1) if anybody has a spare one, that would be great to start.
2) manually run one on AWS
3) auto-create them on AWS as needed (include in trybots too). requires new builder infrastructure (cc @crawshaw)

AWS links:
http://aws.amazon.com/windows/products/ec2/server2003/ -- until July 14, 2015
http://aws.amazon.com/windows/products/ec2/server2003/access/ -- beyond July 14, 2015

/cc @alexbrainman @minux

**Comment by alexbrainman** [comment]
@bradfitz I am not sure Windows Server 2003 support is important. It is effectively Windows XP. Not many people run that OS anymore (except myself). Microsoft discontinue its support soon.

Alex

**Comment by bradfitz** [comment]
We should have a builder for anything we pretend to halfway support.

If we don't support Go for Windows XP/2003, then why is #10162 still open? And the downloads page says Windows XP or above.

**Comment by alexbrainman** [comment]
Fair enough.

**Comment by kardianos** [comment]
@bradfitz I could host them for now. Want both a 386 and amd64 builder? I could also transfer you the ISOs and SKUs; I don't have a need for them aside from something like this now.

**Comment by bradfitz** [comment]
I think just amd64 or just 386 is fine. Whatever is easiest.

I'd prefer you just run one for now. Once I start to run one, I'd rather run it on AWS.

Email me at my golang.org address and I'll get you a builder key.

**Comment by minux** [comment]
@alexbrainman, whether the vendor supports it or not is irrelevant
for this issue as we're still claiming to support XP or above.

We still haven't remove OS X 10.6 support, so we continue to have a
10.6 builder even though we've talked about removing 10.6 support.

We haven't even talked about removing XP/2003 support, and it seems
there is little reason to remove their support.

Modern windows version are using too much resources, so the only
way I can do some windows testing is on windows xp 32-bit and 2003
64-bit.

**Comment by alexbrainman** [comment]
@minux I am not against more builders. I just don't think Windows Server 2003 is a good choice. We still don't have 386 builder (any). And even if you want something that is "very old", Windows XP 386 would me better choice. Windows XP 386 is much more common then Windows Server 2003 amd64. Just my 2c.

Alex

**Comment by minux** [comment]
If you see my replies to the windows 2012 builder thread, I'm actually
in favor of setting up a real 32-bit windows builder (i.e. windows xp).

I suggest that we need a 2003 builder because it's the lowest 64-bit
windows version that we claim to support and it's more common than
64-bit windows xp.

**Comment by kardianos** [comment]
I ran XP 64-bit  for a long time. My understanding is XP 64-bit based off
of 2003 64-bit. In my experience, the architecture matters much more the
the flavor in this case; sometimes the 64-bit version of XP/2003 is a
little "odd".

I think everyone is fine/wants a builder for amd64 and 386. I would suggest
the XP vs 2003 flavor won't matter for the builder's tests.
-Daniel

**Comment by bradfitz** [comment]
WinXP 32-bit would be good too, of course.

**Comment by kardianos** [comment]
I've started the builder. There are currently three failures at tip:
crypto/tls: Missing cert. Should I add from a different box or modify the
test? OS is up to date with MS provided patches.
runtime: The test failure is addressed in my posted CL, pending Alex's
investigation.
time: Several tests fail the accuracy tests; tickers and sleeps wake up too
soon from sleep. I'm honestly not sure if this is because I'm running it in
a VM or if Windows 2003 is just this loose with time (I mean I know its
loose, just a matter of degree).

-Daniel

**Comment by kardianos** [comment]
Also runtime/pprof sometimes fails. I haven't looked into that one.

**Comment by bradfitz** [comment]
We shouldn't install a cert from another box. We need to modify the code and/or test to work on a stock Windows system without modications.

As for missed ticks: first determine whether it's a VM thing. If it's just a VM thing, let's add an environment variable (IN_VM? like the IN_QEMU one we've used before) which, if enabled, skips such tests.

**Comment by bradfitz** [comment]
I have the start of a Windows XP SP3 builder now running on VMWare.

All tests passed with CGO_ENABLED=0 at tip just now. I haven't installed a C compiler yet.

Thanks again to @kardianos for the help.

**Comment by bradfitz** [comment]
I actually worked on the Windows XP builder during my leave a bit. I don't have it online yet, though.

**Comment by bradfitz** [comment]
The XP SP3 builder did a few commits, but failed with:

```
ok  	math/rand	1.441s
ok  	mime	0.146s
ok  	mime/multipart	0.925s
ok  	mime/quotedprintable	0.191s
--- FAIL: TestUDPConnSpecificMethods (0.02s)
	protoconn_test.go:142: write udp4 127.0.0.1:1138->127.0.0.1:1138: wsasendmsg: An invalid argument was supplied.
--- FAIL: TestWriteToUDP (0.00s)
	udpsock_test.go:168: write udp 127.0.0.1:1282: wsasendmsg: An invalid argument was supplied.
FAIL
FAIL	net	10.336s
2017/12/08 15:29:53 Failed: exit status 1
```

... including after I turned off the Windows firewall.

@alexbrainman, do you not also see that failure on your XP machine?

/cc @mikioh

**Comment by alexbrainman** [comment]
> @alexbrainman, do you not also see that failure on your XP machine?

I do see that failure on my Windows XP - I was meant to report / investigate this but I did not have time yet. This is new functionality that has been implemented in CL 76393 and related. The failure is because WSASendMsg returns WSAEINVAL, from https://msdn.microsoft.com/en-us/library/windows/desktop/ms741692(v=vs.85).aspx "The socket has not been bound with bind, or the socket was not created with the overlapped flag.". But on the bottom of that page it also says "Minimum supported client | Windows 8.1, ...". So I doubt this is designed to work on Windows XP. I propose we create an issue for it, and skip broken part of the test.

Alex
